### ISSUE TITLE
yarn 2 Support

### ISSUE THREAD

**Issue by Jordan-Hall** [nrwl/nx#2386]
_[Please make sure you have read the submission guidelines before posting an issue](https://github.com/nrwl/nx/blob/master/CONTRIBUTING.md#-submitting-issue)_

# Prerequisites
Yarn version 2.0.0-rc.27

## Expected Behavior

Should work correctly like before

## Current Behavior

Build failure

## Failure Information (for bugs)

**An unhandled exception occurred: Cannot locate the 'node_modules' directory.
See "AppData\Local\Temp\ng-mO3oFO\angular-errors.log" for further details.**

### Steps to Reproduce

Please provide detailed steps for reproducing the issue.

1. yarn dlx @yarnpkg/doctor .
2. yarn install
3. yarn start

### Context

Yarn has finally released a usable version two. If you using webpack 4 you need to enable PlugNPlay which is native to webpack 5

https://yarnpkg.com/advanced/migration
https://yarnpkg.com/advanced/pnpapi

### Failure Logs

```
[error] Error: Cannot locate the 'node_modules' directory.
    at NgccProcessor.findNodeModulesDirectory (C:\dev\lec-migration\.yarn\$$virtual\@ngtools-webpack-virtual-9bdca6ad48\0\cache\@ngtools-webpack-npm-9.0.0-rc.10-6bfaf96715-1.zip\node_modules\@ngtools\webpack\src\ngcc_processor.js:113:15)
    at new NgccProcessor (C:\dev\lec-migration\.yarn\$$virtual\@ngtools-webpack-virtual-9bdca6ad48\0\cache\@ngtools-webpack-npm-9.0.0-rc.10-6bfaf96715-1.zip\node_modules\@ngtools\webpack\src\ngcc_processor.js:31:43)
    at C:\dev\lec-migration\.yarn\$$virtual\@ngtools-webpack-virtual-9bdca6ad48\0\cache\@ngtools-webpack-npm-9.0.0-rc.10-6bfaf96715-1.zip\node_modules\@ngtools\webpack\src\angular_compiler_plugin.js:554:33
    at SyncHook.eval [as call] (eval at create (C:\dev\lec-migration\.yarn\cache\tapable-npm-1.1.3-f1c2843426-1.zip\node_modules\tapable\lib\HookCodeFactory.js:19:10), <anonymous>:7:1)
    at SyncHook.lazyCompileHook (C:\dev\lec-migration\.yarn\cache\tapable-npm-1.1.3-f1c2843426-1.zip\node_modules\tapable\lib\Hook.js:154:20)
    at Object.webpack [as webpackFactory] (C:\dev\lec-migration\.yarn\cache\webpack-npm-4.41.2-efbe0580aa-1.zip\node_modules\webpack\lib\webpack.js:55:30)
    at createWebpack (C:\dev\lec-migration\.yarn\$$virtual\@angular-devkit-build-webpack-virtual-0beaa59de8\0\cache\@angular-devkit-build-webpack-npm-0.900.0-rc.10-7017fe1d56-1.zip\node_modules\@angular-devkit\build-webpack\src\webpack-dev-server\index.js:20:36)
    at Object.runWebpackDevServer (C:\dev\lec-migration\.yarn\$$virtual\@angular-devkit-build-webpack-virtual-0beaa59de8\0\cache\@angular-devkit-build-webpack-npm-0.900.0-rc.10-7017fe1d56-1.zip\node_modules\@angular-devkit\build-webpack\src\webpack-dev-server\index.js:46:12)
    at SwitchMapSubscriber.project (C:\dev\lec-migration\.yarn\$$virtual\@angular-devkit-build-angular-virtual-d79a835d58\0\cache\@angular-devkit-build-angular-npm-0.900.0-rc.10-f83f279b13-1.zip\node_modules\@angular-devkit\build-angular\src\dev-server\index.js:191:32)
    at SwitchMapSubscriber._next (C:\dev\lec-migration\.yarn\cache\rxjs-npm-6.5.3-bc949c0a94-1.zip\node_modules\rxjs\internal\operators\switchMap.js:49:27)
    at SwitchMapSubscriber.Subscriber.next (C:\dev\lec-migration\.yarn\cache\rxjs-npm-6.5.3-bc949c0a94-1.zip\node_modules\rxjs\internal\Subscriber.js:66:18)
    at C:\dev\lec-migration\.yarn\cache\rxjs-npm-6.5.3-bc949c0a94-1.zip\node_modules\rxjs\internal\util\subscribeToPromise.js:7:24
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

## Other

Any other relevant information that will help us help you.

**Comment by Jordan-Hall** [comment]
Angular appears not to be supporting PNP yet. Tracking https://github.com/angular/angular-cli/issues/12465

**Comment by ElbowBaggins** [comment]
According to [this angular-cli comment](https://github.com/angular/angular-cli/issues/12465#issuecomment-590941650) this is no longer exactly the case.

It looks like native PnP mode support is not in yet, but @yarnpkg/pnpify or the node-modules linker strategy should (allegedly) be usable now.

Of course, I have no idea what this means for you guys, maybe this doesn't help you, but it's a thing now.

**Comment by Jordan-Hall** [comment]
@ElbowBaggins  CLI still now working. getting command nx not found

**Comment by Jordan-Hall** [comment]
@FrozenPandaz 

this issue is now relating to nx cli i believe `https://github.com/nrwl/nx/blob/master/packages/cli/lib/init-local.ts`it's looking directly in the node_module folder

**Comment by vaunus** [comment]
I am seeing the same issue here and disappointed NX CLI is breaking builds on yarn2 given it is now at version 2.1. Our project fails to build with default yarn2 settings, i.e. `pnp` enabled, but works as expected in `node-modules` mode.

```
> yarn nx serve

/project-path/.pnp.js:34448
    throw firstError;
    ^

Error: Qualified path resolution failed - none of the candidates can be found on the disk.

Source path: /project-path/node_modules/@nrwl/tao/index.js
``` 

Looks like this does indeed relate to the `cli` looking for `tao` in the `node_modules` directory as @Jordan-Hall indicated. 

Is it possible to get a fix for this any time soon?

**Comment by Jordan-Hall** [comment]
> I am seeing the same issue here and disappointed NX CLI is breaking builds on yarn2 given it is now at version 2.1. Our project fails to build with default yarn2 settings, i.e. `pnp` enabled, but works as expected in `node-modules` mode.
> 
> ```
> > yarn nx serve
> 
> /project-path/.pnp.js:34448
>     throw firstError;
>     ^
> 
> Error: Qualified path resolution failed - none of the candidates can be found on the disk.
> 
> Source path: /project-path/node_modules/@nrwl/tao/index.js
> ```
> 
> Looks like this does indeed relate to the `cli` looking for `tao` in the `node_modules` directory as @Jordan-Hall indicated.
> 
> Is it possible to get a fix for this any time soon?

Its annoying me now but im too ill to look. Ill carry on soon but they appears to be loads of files referencing direct file path grrr

**Comment by PetarKirov** [comment]
@FrozenPandaz @bekos what's the current status of Yarn v2 PnP support?

**Comment by derHodrig** [comment]
Is still in Progress
https://github.com/angular/angular-cli/issues/16980

**Comment by Jordan-Hall** [comment]
@derHodrig  this is internal of NX thats the issue atm

**Comment by fguitton** [comment]
This is still of interest for ongoing projects

**Comment by hydrosquall** [comment]
> https://github.com/nrwl/nx/blob/master/packages/cli/lib/init-local.ts

It looks like this step was solved in https://github.com/nrwl/nx/pull/3709

Is there a checklist of any other known current issues besides references to the string `node_modules`? After excluding test files (`.spec.ts`), it looks like there  are ~79 references to node_modules (some of them strings) spread across several files remaining.

https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/nrwl/nx%24+node_modules+lang:typescript++-file:.*.spec%5C.ts&patternType=literal

**Comment by KyleBastien** [comment]
We've been running an Nx repo with Yarn v2 PNP on loose mode for a few months now, and _so far_ most things have worked for us.

The biggest exception is the `nx migrate latest` script, which just outright fails on Yarn v2 PNP. My workaround so far has been to switch Yarn v2 to use the `node_modules` folder, run the migrate latest script, and then switch back to using PNP before committing the result.

I would _assume_ the migrate latest script's problem stem from _something_ either not declaring a dependency correctly, or there being a string referencing a node_modules path. I'm just not sure which out of the above ~79 it might be.

**Comment by nemonemi** [comment]
@FrozenPandaz @bekos what's the current status of Yarn v2 PnP support?

**Comment by srosato** [comment]
I wonder how you made it work in loose mode, I've added `pnpMode: loose` to my `.yarnrc` file, re-ran `yarn install`, but I still get an error when running

command: `yarn nx affected:test`

error:

```
(node:45) [MODULE_NOT_FOUND] Error: @nrwl/devkit tried to access rxjs, but it isn't declared in its dependencies; this makes the require call ambiguous and unsound.
```

**Comment by troncali** [comment]
I just converted a template NestJS project I'm working on to an Nx workspace with Yarn PnP enabled, and I seem to have been successful with `@nrwl/nx@12.3.4` so far without needing `pnpMode: loose`.  (Haven't tested all the functionality yet; only building and serving the NestJS backend.)

Adding a `webpack.config.js` file to the NestJS folder worked for me.  It uses the `webpack-pnp-externals` package to tell webpack not to bundle the PnP dependencies.  

```js
/* webpack.config.js */
const PnpWebpackPlugin = require("pnp-webpack-plugin");
const { WebpackPnpExternals } = require("webpack-pnp-externals");

module.exports = (config) => {
	config.resolve.plugins.push(PnpWebpackPlugin);
	config.resolve.extensions.push(".json");
	config.resolveLoader = { plugins: [PnpWebpackPlugin.moduleLoader(module)] };
	config.externals = [WebpackPnpExternals()];
	return config;
};
```

Then update the `build` target in your `workspace.json` file to include the following:

```json
"build": {
  "options": {
    "webpackConfig": "[YOUR/PATH]/webpack.config.js"
  }
}
```

This made `yarn nx serve` work, which I was having trouble with even with `pnpMode: loose` and populating `packageExtensions` in the `.yarnrc.yml` file.

If you want to see an example, the work-in-progress is available on the [dev-nx branch at troncali/nest-vue](https://github.com/troncali/nest-vue/tree/dev-nx).  You can use `yarn nx serve` without the docker containers, and the backend will be available at http://localhost:3001/v1, but note that there is [some minimal setup](https://github.com/troncali/nest-vue/blob/main/guides/setup.md) (need to rename the .env-template file to .env and create two files for the database username and password).

**Comment by KyleBastien** [comment]
> @KyleBastien I wonder how you made it work in loose mode, I've added `pnpMode: loose` to my `.yarnrc` file, re-ran `yarn install`, but I still get an error when running
> 
> command: `yarn nx affected:test`
> 
> error:
> 
> ```
> (node:45) [MODULE_NOT_FOUND] Error: @nrwl/devkit tried to access rxjs, but it isn't declared in its dependencies; this makes the require call ambiguous and unsound.
> ```

You can fix this by utilizing `packageExtensions` in your `.yarnrc.yml` file. Mine currently looks like:

```
packageExtensions:
  '@nrwl/tao@*':
    dependencies:
      yargs: '15.4.1'
  '@nrwl/jest@*':
    dependencies:
      identity-obj-proxy: '3.0.0'
  '@nrwl/devkit@*':
    dependencies:
      rxjs: 6.6.6
  babel-plugin-const-enum@*:
    dependencies:
      '@babel/generator': ^7.9.6
```

Note: These should probably be added to the package.json's of the respective `@nrwl` packages. But Yarn does give you the tools to declare the dependencies yourself, so you're unblocked.

**Comment by srosato** [comment]
Thank you, trying it out, although I agree that we should not have to do this

**Comment by marckassay** [comment]
> My workaround so far has been to switch Yarn v2 to use the `node_modules` folder, run the migrate latest script, and then switch back to using PNP before committing the result.

@KyleBastien You mentioned "switch", what exactly do you mean? I assume setting `nodeLinker: node-modules` in `.yarnrc.yml`, but anything else?

**Comment by KyleBastien** [comment]
@marckassay Here's the steps I follow to get this to work:

1. Remove `pnpMode: loose` from `.yarnrc.yml`
2. Add `nodeLinker: node-modules` to `.yarnrc.yml`
3. Run `yarn` to install all packages in `node_modules`
4. Run `node ./node_modules/@nrwl/tao/index.js migrate latest`
5. Run `yarn` again to install any changed that the migrate script made to `package.json`
6. Run `node ./node_modules/@nrwl/tao/index.js migrate --run-migrations`
7. Remove the `migrations.json` file
8. Remove `nodeLinker: node-modules` from `.yarnrc.yml`
9. Add `pnpMode: loose` to `.yarnrc.yml`
10. Run `yarn` again to remove the `node_modules` directory, and add back the `.pnp.cjs` file

**Comment by Lonli-Lokli** [comment]
Yarn 3 is already available...

**Comment by KyleBastien** [comment]
Yeap, Yarn v3 is unfortunately busted out of the box with an Nx repo. You can see the issue on the Yarn repo here: https://github.com/yarnpkg/berry/issues/3117.

You can use the `link:` protocol in Yarn to workaround this issue. By adding something like this to your `package.json`:

```json
"@name/your-library": "link:./packages/your-library/src",
```

Followed by running `yarn`. This is an unfortunate workaround, because you have to now maintain all your libraries in your `package.json`. A custom generator could probably be made to help automate this process.

This issue is caused by this: https://github.com/microsoft/TypeScript/pull/35206#discussion_r683139533 `getPnpPath` function in the patched version of TypeScript that Yarn v3 uses. Hopefully once this is fixed, `references` will work again with Yarn v3, and the `link:` workaround won't be necessary.

**Comment by benpsnyder** [comment]
Y'all might be interested in [feat(core): implement Yarn v3](https://github.com/nrwl/nx/pull/7918)

**Comment by sledorze** [comment]
@benpsnyder when is this expected to happen?

**Comment by ryanyue123** [comment]
> @marckassay Here's the steps I follow to get this to work:
> 
> 1. Remove `pnpMode: loose` from `.yarnrc.yml`
> 2. Add `nodeLinker: node-modules` to `.yarnrc.yml`
> 3. Run `yarn` to install all packages in `node_modules`
> 4. Run `node ./node_modules/@nrwl/tao/index.js migrate latest`
> 5. Run `yarn` again to install any changed that the migrate script made to `package.json`
> 6. Run `node ./node_modules/@nrwl/tao/index.js migrate --run-migrations`
> 7. Remove the `migrations.json` file
> 8. Remove `nodeLinker: node-modules` from `.yarnrc.yml`
> 9. Add `pnpMode: loose` to `.yarnrc.yml`
> 10. Run `yarn` again to remove the `node_modules` directory, and add back the `.pnp.cjs` file

i was able to get it working by following these steps, but i noticed a lot of the default `nrwl` executors did not work (ex: @nrwl/jest, @nrwl/lint). I just had to default to running `jest --config ./myconfig`. were you able to find a solution for this @KyleBastien ?

**Comment by KyleBastien** [comment]
@ryanyue123 I think that Nx has recently added features for some of these executors that uses [swc](https://github.com/swc-project/swc). Since swc runs native code, it requires an install to trigger downloading that native code associated with the correct platform. I haven't figured out away around this besides running `yarn` (not zero-install) to ensure those native bits get downloaded.

Maybe @benpsnyder can give an update on their PR to implement first class support for yarn v3, that would probably be the "right" fix to this, but I'm not sure if that PR support zero-install or not.

I think at least one change that _could_ be made here would be to move this `if` statement: https://github.com/nrwl/nx/blob/541d4e241bacc8097842fcff9cc8c5c5a743552a/packages/eslint-plugin-nx/src/resolve-workspace-rules.ts#L40

to here: https://github.com/nrwl/nx/blob/bafabb3a11c1d9084e114569f700cd793aac199f/packages/eslint-plugin-nx/src/index.ts#L30

Which would prevent the `resolve-workspace-rules.ts` file from importing from `@swc-node/register/register` unless workspace lint rules are actually used. This would solve this problem for the ESLint plugin at least in the case that you don't use workspace lint rules.

**Comment by ryanyue123** [comment]
> @ryanyue123 I think that Nx has recently added features for some of these executors that uses [swc](https://github.com/swc-project/swc). Since swc runs native code, it requires an install to trigger downloading that native code associated with the correct platform. I haven't figured out away around this besides running `yarn` (not zero-install) to ensure those native bits get downloaded, into the unplugged folder.
> 
> Maybe @benpsnyder can give an update on their PR to implement first class support for yarn v3, that would probably be the "right" fix to this, but I'm not sure if that PR supports zero-install or not.
> 
> I think at least one change that _could_ be made here would be to move this `if` statement:
> 
> https://github.com/nrwl/nx/blob/541d4e241bacc8097842fcff9cc8c5c5a743552a/packages/eslint-plugin-nx/src/resolve-workspace-rules.ts#L40
> 
> to here:
> 
> https://github.com/nrwl/nx/blob/bafabb3a11c1d9084e114569f700cd793aac199f/packages/eslint-plugin-nx/src/index.ts#L30
> 
> Which would prevent the `resolve-workspace-rules.ts` file from importing from `@swc-node/register/register` unless workspace lint rules are actually used. This would solve this problem for the ESLint plugin at least in the case that you don't use workspace lint rules.

Thanks for the reply- i'll probably stick to running my own commands for now until nx officially supports yarn3. hopefully it's soon🤞🏻🤞🏻🤞🏻🤞🏻

**Comment by BigForNothing** [comment]
When running migrate, the command fails due to trying to use the `-W` flag from yarn v1

```
yarn nx migrate latest
Fetching meta data about packages.
It may take a few minutes.
Fetching @nrwl/workspace@latest

 ERROR  The migrate command failed.

Command failed: yarn add -W @nrwl/workspace@latest
```
In yarn v3 (and i believe v2), `-W` isn't used.

**Comment by MrChrisRodriguez** [comment]
Has anyone managed to get globally installed yarn to recognize Nx modules in PnP mode? The local `yarn nx` works great, but I'm curious if we can get global nx working in concert, too.

**Comment by nemonemi** [comment]
@MrChrisRodriguez how did you install NX globally?  As far as I remember, Yarn doesn't support global packages.

**Comment by MrChrisRodriguez** [comment]
@nemonemi iirc both Yarn and Nx recommend using npm to install them globally.

1. npm i -g yarn
2. npm i -g nx

Both Yarn and Nx have mechanisms for checking if your cwd is configured for them and inherits that configuration. Global Nx hasn't implemented the PnP API, though, so you have to call it within the yarn context `yarn nx ...` or do what I did and just create the alias: `alias ynx='yarn nx'`

**Comment by giovai** [comment]
Does Nx 12.9 support Yarn 3 PnP?

**Comment by ild0tt0re** [comment]
I'm trying to switch to **yarn 3.2.1 with PnP mode** but despite having installed the missing package `next-compose-plugins`, it keeps giving me the same error:
```
yarn run nx serve hanami

> nx run hanami:serve:development

info  - Loaded env from apps/hanami/.env.development
info  - Loaded env from apps/hanami/.env
error - Failed to load next.config.js, see more info here https://nextjs.org/docs/messages/next-config-error
Your application tried to access next-compose-plugins, but it isn't declared in your dependencies; this makes the require call ambiguous and unsound.

Required package: next-compose-plugins
Required by: /Users/angeloca/dev/fe.monorepo/apps/hanami/

Require stack:
- /Users/angeloca/dev/fe.monorepo/apps/hanami/next.config.js
```

**Comment by Lonli-Lokli** [comment]
Is there any work in progress from Nx team?
Angular seems like supports almost everything while task is still not closed

**Comment by ambeur** [comment]
Looking forward to this feature too!

**Comment by Lonli-Lokli** [comment]
> **Note: All of the `@nrwl` package extensions here are unnecessary after Nx 12.6.2, because they were added to the respective package.json's**

@KyleBastien @JamesHenry I got same error **without** Nx with just angular-eslint. Should it be fixed there or here?
```
Required package: @angular-devkit/architect (via "@angular-devkit\architect")
Required by: @nrwl/devkit@npm:13.1.3 (via D:\gitlab\cv-app\frontend\cv-ui\.yarn\cache\@nrwl-devkit-npm-13.1.3-a349b36d70-1cfd19866f.zip\node_modules\@nrwl\devkit\src\utils\)

Require stack:
- D:\gitlab\cv-app\frontend\cv-ui\.yarn\cache\@nrwl-devkit-npm-13.1.3-a349b36d70-1cfd19866f.zip\node_modules\@nrwl\devkit\src\utils\convert-nx-executor.js
- D:\gitlab\cv-app\frontend\cv-ui\.yarn\cache\@nrwl-devkit-npm-13.1.3-a349b36d70-1cfd19866f.zip\node_modules\@nrwl\devkit\index.js
- D:\gitlab\cv-app\frontend\cv-ui\.yarn\__virtual__\@angular-eslint-builder-virtual-c8ba735627\0\cache\@angular-eslint-builder-npm-14.0.0-93f7fe6dc0-27e23a86bf.zip\node_modules\@angular-eslint\builder\dist\compat.js
- D:\gitlab\cv-app\frontend\cv-ui\.yarn\cache\@angular-devkit-architect-npm-0.1400.5-03eb80d39b-afb57021b2.zip\node_modules\@angular-devkit\architect\node\node-modules-architect-host.js     
- D:\gitlab\cv-app\frontend\cv-ui\.yarn\cache\@angular-devkit-architect-npm-0.1400.5-03eb80d39b-afb57021b2.zip\node_modules\@angular-devkit\architect\node\index.js
- D:\gitlab\cv-app\frontend\cv-ui\.yarn\cache\@angular-cli-npm-14.0.5-eccde1fd20-5bc4ae64b3.zip\node_modules\@angular\cli\src\command-builder\architect-base-command-module.js
- D:\gitlab\cv-app\frontend\cv-ui\.yarn\cache\@angular-cli-npm-14.0.5-eccde1fd20-5bc4ae64b3.zip\node_modules\@angular\cli\src\command-builder\architect-command-module.js
- D:\gitlab\cv-app\frontend\cv-ui\.yarn\cache\@angular-cli-npm-14.0.5-eccde1fd20-5bc4ae64b3.zip\node_modules\@angular\cli\src\commands\build\cli.js
- D:\gitlab\cv-app\frontend\cv-ui\.yarn\cache\@angular-cli-npm-14.0.5-eccde1fd20-5bc4ae64b3.zip\node_modules\@angular\cli\src\command-builder\command-runner.js
- D:\gitlab\cv-app\frontend\cv-ui\.yarn\cache\@angular-cli-npm-14.0.5-eccde1fd20-5bc4ae64b3.zip\node_modules\@angular\cli\lib\cli\index.js
- D:\gitlab\cv-app\frontend\cv-ui\.yarn\cache\@angular-cli-npm-14.0.5-eccde1fd20-5bc4ae64b3.zip\node_modules\@angular\cli\lib\init.js
```

And here is why
```
PS D:\gitlab\cv-app\frontend\cv-ui> yarn why @nrwl/devkit
├─ @angular-eslint/builder@npm:14.0.0
│  └─ @nrwl/devkit@npm:13.1.3 (via npm:13.1.3)
│
└─ @angular-eslint/builder@npm:14.0.0 [f4bce]
   └─ @nrwl/devkit@npm:13.1.3 (via npm:13.1.3)
```

**Comment by JamesHenry** [comment]
@Lonli-Lokli I can't comment about yarn 2/pnp because I am not familiar with it, but this has flagged up that I forgot to update the dependency on `@nrwl/devkit` for `angular-eslint` v14 so thank you for raising that, I will release a patch today

**Comment by snebjorn** [comment]
I think `@nrwl/devkit` also need to fix their dependency to `@angular-devkit/architect` as I'm getting this error when trying to lint an angular project using `@angular-eslint@14.0.2`.

```
❯ yarn lint
An unhandled exception occurred: @nrwl/devkit tried to access @angular-devkit/architect, but it isn't declared in its dependencies; this makes the require call ambiguous and unsound.

Required package: @angular-devkit/architect (via "@angular-devkit\architect")
Required by: @nrwl/devkit@virtual:15288c7928c7c18c53a14dcfc084d0ce97806beb2231966887b0044d44e6502472ae028c1349166fbbbb635fc5f5cd532021a516d101259b0db620cd6b2b98a9#npm:14.4.2 (via D:\Users\dk8EskDi\Source\ng14-pnpm\.yarn\__virtual__\@nrwl-devkit-virtual-824b188259\0\cache\@nrwl-devkit-npm-14.4.2-4aa707a2fe-ed4c619803.zip\node_modules\@nrwl\devkit\src\utils\)
```

```
❯ yarn why @nrwl/devkit 
├─ @angular-eslint/builder@npm:14.0.2
│  └─ @nrwl/devkit@npm:14.4.2 (via npm:^14.2.4)
│
└─ @angular-eslint/builder@npm:14.0.2 [78dc4]
   └─ @nrwl/devkit@npm:14.4.2 [15288] (via npm:^14.2.4 [15288])
```

**Comment by MalikBagwala** [comment]
@AgentEnder @FrozenPandaz Any updates on this? this feature will make CI jobs very fast and efficient since you dont need to run yarn install on any step (which is a pain considering the number of packages in a mono-repo)

**Comment by tyteen4a03** [comment]
+1 - yarn 1 has been discontinued for 3 years now.

**Comment by muratbeser** [comment]
https://github.com/angular/angular-cli/issues/16980

**Comment by blackholegalaxy** [comment]
I got some weird behavior in tests. Lint works, build works, but running jest tests against angular somehow fails on some files.

```
 FAIL   app-core  libs/app-core/src/lib/services/error-handler/error-handler.service.spec.ts
  ● Test suite failed to run

    TypeError: Cannot read properties of undefined (reading 'resolvedFileName')

      at module.exports (../../.yarn/cache/@nrwl-jest-npm-15.2.1-1e6535ae55-e208605944.zip/packages/jest/plugins/resolver.ts:85:22)

 FAIL   app-core  libs/app-core/src/lib/store/effects/router.effect.spec.ts
  ● Test suite failed to run

    TypeError: Cannot read properties of undefined (reading 'resolvedFileName')

      at module.exports (../../.yarn/cache/@nrwl-jest-npm-15.2.1-1e6535ae55-e208605944.zip/packages/jest/plugins/resolver.ts:85:22)

 PASS   app-core  libs/app-core/src/lib/store/actions/app-notification.action.spec.ts (8.93 s)
  App Notification Actions
    add one
      √ should create an action (2 ms)
```

As you can see the tests on some fails. 

Here is the line mentionned in the error:
```ts
 return ts.resolveModuleName(path, options.basedir, compilerOptions, host)
      .resolvedModule.resolvedFileName;
```

https://github.com/nrwl/nx/blob/master/packages/jest/plugins/resolver.ts#L85

Any idea on how to solve these?

Here is my current packageExtensions tree so far:

```yaml
packageExtensions:
  "@nrwl/devkit@*":
    dependencies:
      "@angular-devkit/architect": "*"
  "@angular-eslint/schematics@*":
    dependencies:
      "@angular-devkit/schematics": "*"
      "@angular-devkit/core": "*"
  "@angular-eslint/builder@*":
    dependencies:
      "@angular-devkit/architect": "*"
  "jest-preset-angular@*":
    dependencies:
      "jest-util": "*"
```

**Comment by quicksilverr** [comment]
Hey, any updates regarding this? 
Does yarn 2 now work with nx?

**Comment by Peeja** [comment]
To be clear, as I understand it, Nx absolutely supports Yarn 2+, just not PnP mode. PnP is one of the main developments in Yarn 2, but the traditional `node_modules` mode is still very much supported.

It would be fantastic to have PnP support in Nx, but it's not as dire as only supporting Yarn 1, which as @tyteen4a03 says is no longer in active development.

**Comment by tyteen4a03** [comment]
@Peeja I struggled to get Yarn2 to work even in node linking mode.

**Comment by Peeja** [comment]
@tyteen4a03 Ah, *that* seems more concerning. I haven't tried myself in some time, so I was going by the comments above. But I wonder if conflating the issues is part of what keeps this from moving forward.

**Comment by AgentEnder** [comment]
iirc yarn2 with node modules linker should absolutely work at this point. If anyone is actively experiencing issues in latest please open a new report.

**Comment by slikts** [comment]
When getting used to the zero-installs workflow with Yarn PnP, switching back to `node_modules` is a bit painful, so it'd be really nice to have this feature.

**Comment by ild0tt0re** [comment]
Is there any roadmap on the work required to implement this feature?

**Comment by doughlass** [comment]
> Is there any roadmap on the work required to implement this feature?

Apparently not.

**Comment by vire** [comment]
FYI `16.2.1` works for me with yarn2 - tho I'm using own `webpack` via run command, but then the next in 16.3.2 is broken as it calls `npx next build`

**Comment by amartincolby** [comment]
So after reading through this, I assume that the conclusion is, Yarn PnP is not supported and there is no timeframe in which it will be supported? I am willing to transition to Pnpm in service of switching to Nx, but are there at least plans to make a future transition to PnP easy?

**Comment by meeroslav** [comment]
I'm happy to announce that we are working on the PnP support. A lot of the issues have been resolved already (you can try it out with the latest [beta](https://github.com/nrwl/nx/releases/tag/16.6.0-beta.7)).

There are two known issues that we need to fix:
- Swc compiler fix from @nx/js was fixed but didn't make it on time for the beta cut
- NextJS build fix was reverted due to some issues with our own docs app

If you can, try this beta and let us know if you find any issues (apart from those listed above).

**Comment by sushruth** [comment]
@meeroslav this is the best news I have heard all year. Yarn pnp is not perfect but it is a great option to choose in the current JS ecosystem.

**Comment by amartincolby** [comment]
This could not have come at a better time. I was just balancing pure Yarn Workspaces vs Nx specifically because of this issue.

**Comment by meeroslav** [comment]
Swc compiler is fixed in the latest [16.6.0-beta.8](https://github.com/nrwl/nx/releases/tag/16.6.0-beta.8)

**Comment by amartincolby** [comment]
How do we enable Plug-n-Play? If I generate a workspace with yarn v3, it successfully uses yarn Berry but does not generate a .pnp.cjs file. It still defaults to using nodeLinker:node-modules.

**Comment by meeroslav** [comment]
You can use this reference:
https://nx.dev/recipes/tips-n-tricks/yarn-pnp

We do not provide a flag for create-nx-workspace to enable PnP automatically.

**Comment by meeroslav** [comment]
Currently, our `@nx/angular` package is the only one not supporting `PnP` due to third-party blockers.

If you detect any issues with any other package, please report it here.

**Comment by meeroslav** [comment]
The latest beta has all the changes related to Yarn PnP support. This concludes this feature request.

If you find any problems with Yarn PnP, please open a new issue. Thank you for your patience!
